### ISSUE TITLE
Should Intel's mkl-library be part of a FOSS toolchain?

### ISSUE THREAD

**Issue by cmeesters** [easybuilders/easybuild-easyconfigs#14332]
Hi,

as discussed on mattermost, I - once more - would like to address a point of concern: As of 2021b imkl is part of the FOSS toolchain, because flexiblas has imkl as a mandatory requirement.

Personally, I confess ideological reservations as well as I would like to address the concern that Intel, in the past, has not contributed steadily to the open-source community, but rather with a hefty amount of changes (license paths, ways to install, etc.) from release to release of various softwares. 

What do you think? foss (and hence, flexiblas) with or without imkl?

**Comment by bartoldeman** [comment]
foss should be without imkl. It was just nice to have the option to use imkl with FlexiBLAS.

We made it a builddependency also for that reason so that users that the foss module itself doesn't depend on imkl.

There are some technical solutions to allow using imkl with FlexiBLAS anyway but they'd change either imkl itself or setting up a FlexiBLAS-imkl easyconfig. But removing it from the default FlexiBLAS easyconfig is simple and can be done in parallel.

**Comment by bartoldeman** [comment]
For MKL it can produce a FlexiBLAS compatible backend independently via
https://github.com/mpimd-csc/flexiblas/issues/13#issuecomment-782828653
the question is then mostly how to teach FlexiBLAS about the backends.
There are a few methods at least
1. `$FLEXIBLAS_LIBRARY_PATH`
2. let flexiblas add a conf file with IMKL and the location where it would be installed, without needing to depend on IMKL. This is tricky because EB doesn't know where IMKL would be installed without the module actually being there.
3. A combination: the conf file e.g. `imk.conf` installed by FlexiBLAS in `$EBROOTFLEXIBLAS/etc/flexiblasrc.d` has
```
[imkl]
library = libflexiblas_imkl.so
```
which can then be found via `FLEXIBLAS_LIBRARY_PATH` iff the `imkl` module is loaded.

**Comment by boegel** [comment]
A `FlexiBLAS-imkl` easyconfig is probably the best way forward here...

Does the `imkl.conf` for FlexiBLAS need to be a part of the FlexiBLAS installation, or can you specify additional paths to consider for FlexiBLAS configuration files?

**Comment by bartoldeman** [comment]
imkl.conf needs to a part of the FlexiBLAS installation. There is a `FLEXIBLAS_CONFIG` environment variable but that overrides everything, it's not a priority list of config files / directories, unlike FLEXIBLAS_LIBRARY_PATH.

If you want to be consistent you can also build FlexiBLAS with zero backends (apart from netlib), and have it create a config file with just the name of the libraries, something like:
```
[imkl]
library = libflexiblas_imkl.so
[BLIS]
library = libblis.so
[OpenBLAS]
library = libopenblas.so
```
then in combination with `FLEXIBLAS_LIBRARY_PATH` (set by imkl/BLIS/OpenBLAS modules) it can do the job.
